runtime/parachains/src/runtime_api_impl/vstaging.rs

```diff
@@ -26,6 +29,48 @@ pub fn get_session_disputes<T: disputes::Config>(
 	<disputes::Pallet<T>>::disputes()
 }
 
-// TODO [now]: implicit `validity_constraints`. Ensure that `min_relay_parent`
-// never goes lower than the point at which asynchronous backing was enabled.
-// Also, never cross session boundaries.
+/// Implementation for `StagingValidityConstraints` function from the runtime API
+pub fn validity_constraints<T: initializer::Config>(
+	para_id: ParaId,
+) -> Option<Constraints<T::BlockNumber>> {
+	// Async backing is only expected to be enabled with a tracker capacity of 1.
+	// Subsequent configuration update gets applied on new session, which always
+	// clears the buffer.
+	//
+	// Thus, minimum relay parent is ensured to have asynchronous backing enabled.
+	let min_relay_parent_number =
+		<shared::Pallet<T>>::allowed_relay_parents().earliest_block_number()?;
+	let required_parent = <paras::Pallet<T>>::para_head(para_id)?;
+	let validation_code_hash = <paras::Pallet<T>>::current_code_hash(para_id)?;
+
+	let upgrade_restriction = <paras::Pallet<T>>::upgrade_restriction_signal(para_id);
+	let future_validation_code =
+		<paras::Pallet<T>>::future_code_upgrade_at(para_id).and_then(|block_num| {
+			// Only read the storage if there's a pending upgrade.
+			Some(block_num).zip(<paras::Pallet<T>>::future_code_hash(para_id))
+		});
+
+	let dmp_remaining_messages = <dmp::Pallet<T>>::dmq_length(para_id);
+
+	let (ump_remaining, ump_remaining_bytes) = <ump::Pallet<T>>::relay_dispatch_queue_size(para_id);
+
+	let hrmp_inbound = <hrmp::Pallet<T>>::inbound_hrmp_constraints(para_id);
+	let hrmp_channels_out = <hrmp::Pallet<T>>::outbound_hrmp_constraints(para_id);
+
+	let config = <configuration::Pallet<T>>::config();
+	Some(Constraints {
+		min_relay_parent_number,
+		max_pov_size: config.max_pov_size,
+		max_code_size: config.max_code_size,
+		ump_remaining,
+		ump_remaining_bytes,
+		dmp_remaining_messages,
```

Comment:
Any way to check an advancement rule in the inclusion emulator?

Reply:
This is the advancement rule:

https://github.com/paritytech/polkadot/blob/21e19328bd8d3130acf50394de2ec4c19cdcaa67/runtime/parachains/src/dmp.rs#L184-L191

So the inclusion emulator here should be able to do the same check here: https://github.com/paritytech/polkadot/blob/21e19328bd8d3130acf50394de2ec4c19cdcaa67/node/subsystem-util/src/inclusion_emulator/staging.rs#L313-L318

Reply:
Meaning, it's also missing and I should add it too?

Reply:
https://github.com/paritytech/polkadot/issues/6267 I filed this. Nice catch